File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/access/LibraryRepository.java
package com.lasthopesoftware.bluewater.client.library.access;

import android.content.Context;
import android.database.SQLException;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.repository.CloseableTransaction;
import com.lasthopesoftware.bluewater.repository.InsertBuilder;
import com.lasthopesoftware.bluewater.repository.RepositoryAccessHelper;
import com.lasthopesoftware.bluewater.repository.UpdateBuilder;
import com.namehillsoftware.artful.Artful;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.queued.MessageWriter;
import com.namehillsoftware.handoff.promises.queued.QueuedPromise;
import com.namehillsoftware.lazyj.Lazy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collection;

public class LibraryRepository implements ILibraryStorage, ILibraryProvider {
	private final Context context;

	public LibraryRepository(Context context) {
		this.context = context;
	}

	@Override
	public Promise<Library> getLibrary(int libraryId) {
		return new QueuedPromise<>(new GetLibraryWriter(context, libraryId), RepositoryAccessHelper.databaseExecutor);
	}

	@Override
	public Promise<Collection<Library>> getAllLibraries() {
		return new QueuedPromise<>(new GetAllLibrariesWriter(context), RepositoryAccessHelper.databaseExecutor);
	}

	@Override
	public Promise<Library> saveLibrary(Library library) {
		return new QueuedPromise<>(new SaveLibraryWriter(context, library), RepositoryAccessHelper.databaseExecutor);
	}

	private static class GetAllLibrariesWriter implements MessageWriter<Collection<Library>> {

		private Context context;

		private GetAllLibrariesWriter(Context context) {
			this.context = context;
		}

		@Override
		public Collection<Library> prepareMessage() {
			try (RepositoryAccessHelper repositoryAccessHelper = new RepositoryAccessHelper(context)) {
				return
					repositoryAccessHelper
						.mapSql("SELECT * FROM " + Library.tableName)
						.fetch(Library.class);
			}
		}
	}

	private static class GetLibraryWriter implements MessageWriter<Library> {

		private int libraryId;
		private Context context;

		private GetLibraryWriter(Context context, int libraryId) {
			this.libraryId = libraryId;
			this.context = context;
		}

		@Override
		public Library prepareMessage() {
			if (libraryId < 0) return null;

			try (RepositoryAccessHelper repositoryAccessHelper = new RepositoryAccessHelper(context)) {
				return
					repositoryAccessHelper
						.mapSql("SELECT * FROM " + Library.tableName + " WHERE id = @id")
						.addParameter("id", libraryId)
						.fetchFirst(Library.class);
			}
		}
	}

	private static class SaveLibraryWriter implements MessageWriter<Library> {

		private static final Logger logger = LoggerFactory.getLogger(SaveLibraryWriter.class);

		private static final Lazy<String> libraryInsertSql
			= new Lazy<>(() ->
				InsertBuilder
					.fromTable(Library.tableName)
					.addColumn(Library.accessCodeColumn)
					.addColumn(Library.authKeyColumn)
					.addColumn(Library.isLocalOnlyColumn)
					.addColumn(Library.libraryNameColumn)
					.addColumn(Library.isRepeatingColumn)
					.addColumn(Library.customSyncedFilesPathColumn)
					.addColumn(Library.isSyncLocalConnectionsOnlyColumn)
					.addColumn(Library.isUsingExistingFilesColumn)
					.addColumn(Library.nowPlayingIdColumn)
					.addColumn(Library.nowPlayingProgressColumn)
					.addColumn(Library.savedTracksStringColumn)
					.addColumn(Library.selectedViewColumn)
					.addColumn(Library.selectedViewTypeColumn)
					.addColumn(Library.syncedFileLocationColumn)
					.build());

		private static final Lazy<String> libraryUpdateSql
			= new Lazy<>(() ->
				UpdateBuilder
					.fromTable(Library.tableName)
					.addSetter(Library.accessCodeColumn)
					.addSetter(Library.authKeyColumn)
					.addSetter(Library.isLocalOnlyColumn)
					.addSetter(Library.libraryNameColumn)
					.addSetter(Library.isRepeatingColumn)
					.addSetter(Library.customSyncedFilesPathColumn)
					.addSetter(Library.isSyncLocalConnectionsOnlyColumn)
					.addSetter(Library.isUsingExistingFilesColumn)
					.addSetter(Library.nowPlayingIdColumn)
					.addSetter(Library.nowPlayingProgressColumn)
					.addSetter(Library.savedTracksStringColumn)
					.addSetter(Library.selectedViewColumn)
					.addSetter(Library.selectedViewTypeColumn)
					.addSetter(Library.syncedFileLocationColumn)
					.setFilter("WHERE id = @id")
					.buildQuery());


		private Context context;
		private Library library;

		private SaveLibraryWriter(Context context, Library library) {
			this.context = context;
			this.library = library;
		}

		@Override
		public Library prepareMessage() {
			try (RepositoryAccessHelper repositoryAccessHelper = new RepositoryAccessHelper(context)) {
				try (CloseableTransaction closeableTransaction = repositoryAccessHelper.beginTransaction()) {
					final boolean isLibraryExists = library.getId() > -1;

					final Artful artful =
						repositoryAccessHelper
							.mapSql(isLibraryExists ? libraryUpdateSql.getObject() : libraryInsertSql.getObject())
							.addParameter(Library.accessCodeColumn, library.getAccessCode())
							.addParameter(Library.authKeyColumn, library.getAuthKey())
							.addParameter(Library.isLocalOnlyColumn, library.isLocalOnly())
							.addParameter(Library.libraryNameColumn, library.getLibraryName())
							.addParameter(Library.isRepeatingColumn, library.isRepeating())
							.addParameter(Library.customSyncedFilesPathColumn, library.getCustomSyncedFilesPath())
							.addParameter(Library.isSyncLocalConnectionsOnlyColumn, library.isSyncLocalConnectionsOnly())
							.addParameter(Library.isUsingExistingFilesColumn, library.isUsingExistingFiles())
							.addParameter(Library.nowPlayingIdColumn, library.getNowPlayingId())
							.addParameter(Library.nowPlayingProgressColumn, library.getNowPlayingProgress())
							.addParameter(Library.savedTracksStringColumn, library.getSavedTracksString())
							.addParameter(Library.selectedViewColumn, library.getSelectedView())
							.addParameter(Library.selectedViewTypeColumn, library.getSelectedViewType())
							.addParameter(Library.syncedFileLocationColumn, library.getSyncedFileLocation());

					if (isLibraryExists)
						artful.addParameter("id", library.getId());

					final long result = artful.execute();
					closeableTransaction.setTransactionSuccessful();

					if (!isLibraryExists)
						library.setId((int) result);

					logger.debug("Library saved.");
					return library;
				} catch (SQLException se) {
					logger.error("There was an error saving the library", se);
					return null;
				}
			}
		}
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/library/access/LibraryRepository.kt
package com.lasthopesoftware.bluewater.client.browsing.library.access

import android.content.Context
import com.lasthopesoftware.bluewater.BuildConfig
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryEntityInformation.isRepeatingColumn
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryEntityInformation.nowPlayingIdColumn
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryEntityInformation.nowPlayingProgressColumn
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryEntityInformation.savedTracksStringColumn
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryEntityInformation.tableName
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryNowPlayingValues
import com.lasthopesoftware.bluewater.repository.RepositoryAccessHelper
import com.lasthopesoftware.bluewater.repository.fetch
import com.lasthopesoftware.bluewater.repository.fetchFirstOrNull
import com.lasthopesoftware.bluewater.repository.insert
import com.lasthopesoftware.bluewater.repository.update
import com.lasthopesoftware.bluewater.shared.lazyLogger
import com.lasthopesoftware.resources.executors.ThreadPools.promiseTableMessage
import com.namehillsoftware.handoff.cancellation.CancellationSignal
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.handoff.promises.queued.cancellation.CancellableMessageWriter
import com.namehillsoftware.querydroid.SqLiteAssistants
import java.io.IOException
import kotlin.coroutines.cancellation.CancellationException

class LibraryRepository(private val context: Context) : ManageLibraries, ProvideLibraries {
	override fun promiseLibrary(libraryId: LibraryId): Promise<Library?> =
		promiseTableMessage(GetLibraryWriter(context, libraryId))

	override fun promiseAllLibraries(): Promise<Collection<Library>> =
		promiseTableMessage(GetAllLibrariesWriter(context))

	override fun saveLibrary(library: Library): Promise<Library> =
		promiseTableMessage(SaveLibraryWriter(context, library))

	override fun updateNowPlaying(values: LibraryNowPlayingValues): Promise<Unit> =
		promiseTableMessage(UpdateNowPlayingLibraryWriter(context, values))

	override fun promiseNowPlayingValues(libraryId: LibraryId): Promise<LibraryNowPlayingValues?> =
		promiseTableMessage(GetNowPlayingValuesWriter(context, libraryId))

	override fun removeLibrary(libraryId: LibraryId): Promise<Unit> =
		promiseTableMessage(RemoveLibraryWriter(context, libraryId))

	private class GetAllLibrariesWriter(private val context: Context) : CancellableMessageWriter<Collection<Library>> {
		override fun prepareMessage(cancellationSignal: CancellationSignal): Collection<Library> {
			if (cancellationSignal.isCancelled) throw CancellationException("Cancelled while getting libraries")
			return RepositoryAccessHelper(context).use { repositoryAccessHelper ->
				repositoryAccessHelper.beginNonExclusiveTransaction().use {
					repositoryAccessHelper
						.mapSql("SELECT * FROM $tableName")
						.fetch()
				}
			}
		}
	}

	private class GetLibraryWriter(private val context: Context, private val libraryId: LibraryId) : CancellableMessageWriter<Library?> {

		override fun prepareMessage(cancellationSignal: CancellationSignal): Library? {
			val libraryInt = libraryId.id
			if (libraryInt < 0) return null
			if (cancellationSignal.isCancelled) throw CancellationException("Cancelled while getting library.")

			return RepositoryAccessHelper(context).use { repositoryAccessHelper ->
				repositoryAccessHelper.beginNonExclusiveTransaction().use {
					repositoryAccessHelper
						.mapSql("SELECT * FROM $tableName WHERE id = @id")
						.addParameter("id", libraryInt)
						.fetchFirstOrNull()
				}
			}
		}
	}

	private class GetNowPlayingValuesWriter(private val context: Context, private val libraryId: LibraryId) : CancellableMessageWriter<LibraryNowPlayingValues?> {

		override fun prepareMessage(cancellationSignal: CancellationSignal): LibraryNowPlayingValues? {
			val libraryInt = libraryId.id
			if (libraryInt < 0) return null
			if (cancellationSignal.isCancelled) throw CancellationException("Cancelled while getting now playing values.")

			return RepositoryAccessHelper(context).use { repositoryAccessHelper ->
				repositoryAccessHelper.beginNonExclusiveTransaction().use {
					repositoryAccessHelper
						.mapSql("SELECT id, $nowPlayingIdColumn, $nowPlayingProgressColumn, $savedTracksStringColumn, $isRepeatingColumn FROM $tableName WHERE id = @id")
						.addParameter("id", libraryInt)
						.fetchFirstOrNull()
				}
			}
		}
	}

	private class SaveLibraryWriter(private val context: Context, private val library: Library) : CancellableMessageWriter<Library> {

		companion object {
			private val logger by lazyLogger<SaveLibraryWriter>()

		}

		override fun prepareMessage(cancellationSignal: CancellationSignal): Library {
			if (cancellationSignal.isCancelled) throw CancellationException("Cancelled while saving library")

			return RepositoryAccessHelper(context).use { repositoryAccessHelper ->
				val isLibraryExists = library.id > -1

				val returnLibrary =
					if (isLibraryExists) repositoryAccessHelper.update(tableName, library)
					else repositoryAccessHelper.insert(tableName, library)

				if (BuildConfig.DEBUG) {
					logger.debug("Library saved.")
				}

				returnLibrary
			}
		}
	}

	private class UpdateNowPlayingLibraryWriter(
		private val context: Context,
		private val values: LibraryNowPlayingValues
	) : CancellableMessageWriter<Unit> {

		companion object {
			private val logger by lazyLogger<SaveLibraryWriter>()
		}

		override fun prepareMessage(cancellationSignal: CancellationSignal) {
			if (cancellationSignal.isCancelled) throw CancellationException("Cancelled while saving now playing values.")

			val libraryInt = values.id
			if (libraryInt < 0) return

			RepositoryAccessHelper(context).use { repositoryAccessHelper ->
				val result = SqLiteAssistants.updateValue(repositoryAccessHelper.writableDatabase, tableName, values)

				if (result == 0L) {
					throw IOException("Updating $tableName for id ${values.id} returned 0 rows.")
				}

				if (BuildConfig.DEBUG) {
					logger.debug("Now Playing updated for library {}.", libraryInt)
				}
			}
		}
	}

	private class RemoveLibraryWriter(private val context: Context, private val libraryId: LibraryId) : CancellableMessageWriter<Unit> {

		override fun prepareMessage(cancellationSignal: CancellationSignal) {
			val libraryInt = libraryId.id
			if (libraryInt < 0 || cancellationSignal.isCancelled) return
			RepositoryAccessHelper(context).use { repositoryAccessHelper ->
				repositoryAccessHelper.beginTransaction().use {
					repositoryAccessHelper
						.mapSql("DELETE FROM $tableName WHERE id = @id")
						.addParameter("id", libraryInt)
						.execute()
					it.setTransactionSuccessful()
				}
			}
		}
	}
}
